src/CargoBuild/CargoBuildTask.cs

```diff
@@ -0,0 +1,437 @@
+            }
+
+            return exitCode;
+        }
+
+        private async Task<ExitCode> DoRegistryAuthAsync(string workingDir)
+        {
+            return await ExecuteWithRetriesAsync(_cargoPath, $"login --registry {RegistryFeedName}", workingDir, retryCount: 2);
+        }
+
+        private async Task<ExitCode> ExecuteProcessAsync(string fileName, string args, string workingDir, Dictionary<string, string>? envars = null)
+        {
+            try
+            {
+                var processTask = new Task<int>(() =>
+                {
+                    var info = new ProcessStartInfo
+                    {
+                        RedirectStandardError = true,
+                        RedirectStandardOutput = true,
+                        UseShellExecute = false,
+                        FileName = fileName,
+                        Arguments = args,
+                        WorkingDirectory = workingDir,
+                    };
+                    if (envars != null && envars.Count > 0)
+                    {
+                        foreach (var envar in envars)
+                        {
+                            if (!info.EnvironmentVariables.ContainsKey(envar.Key))
+                            {
+                                info.EnvironmentVariables.Add(envar.Key, envar.Value);
+                            }
+                        }
+                    }
+
+                    var process = Process.Start(info);
+                    int maxWait = 1000 * 60 * 15;
+
+                    var nameAndExtension = fileName.Substring(fileName.LastIndexOf('\\') + 1);
+                    Log.LogMessage(MessageImportance.Normal, $"\t\t\t\t\t\t*********** Start {nameAndExtension} logs ***********\n\n");
+
+                    using StreamReader errReader = process!.StandardError;
+                    _ = Log.LogMessagesFromStream(errReader, MessageImportance.Normal);
+
+                    using StreamReader outReader = process!.StandardOutput;
+                    _ = Log.LogMessagesFromStream(outReader, MessageImportance.Normal);
+                    Log.LogMessage(MessageImportance.Normal, $"\t\t\t\t\t\t*********** End {nameAndExtension} logs ***********\n\n");
+                    bool exited = process.WaitForExit(maxWait);
+                    if (!exited)
+                    {
+                        process.Kill();
+                        Log.LogError($"Killed process after max timeout reached : '{info.FileName}'");
+                        return -1;
+                    }
+
+                    return process.ExitCode;
+                });
+
+                processTask.Start();
+                return (ExitCode)await processTask;
+            }
+            catch (Exception ex)
+            {
+                Log.LogWarningFromException(ex);
+                return await Tasks.Task.FromResult(ExitCode.Failed);
+            }
+        }
+
+        private async Task<bool> DownloadRustupAsync()
```

Comment:
Shouldn't this be attached to the Restore? Network Isolation / Firewall requirements forbid the build itself from hitting the network.